bpo-41279: Add StreamReaderBufferedProtocol

tontinton opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-41279](https://bugs.python.org/issue41279) -->
https://bugs.python.org/issue41279
<!-- /issue-number -->

I got way better performance on ``await reader.read()`` using this branch on linux (check out the chart on the ``server.py`` script's comments).

The way I tested was writing a server / client:
``server.py``:

```python
import asyncio
import contextlib
import time


async def client_connected(reader, writer):
    start = time.time()

    with contextlib.closing(writer):
        for i in range(1000):
            # the more this parameter's distance from 65536 is greater the better the performance global_buffer gives
            # On linux:
            # 65536 * 2 -> Gives about 160% better performance
            # 4096 -> Gives about 155% better performance
            # 65536 -> Gives about 125% better performance

            # On windows using 65536 gives the same performance for some reason, which is interesting
            # But any other value gives a bit better performance, for example 65536 * 2 gives about 120% better performance
            await reader.read(65536 * 2)

    print(f'{time.time() - start}')


async def main():
    server = await asyncio.start_server(client_connected, '127.0.0.1', 8888, global_buffer=True)

    addr = server.sockets[0].getsockname()
    print(f'Serving on {addr}')

    async with server:
        await server.serve_forever()


if __name__ == "__main__":
    asyncio.run(main())
```

``client.py``:
```python
import asyncio
import contextlib


async def flood(ip, port):
    message = b'A' * 1024 * 64  # tweak this parameter as much as you want
    reader, writer = await asyncio.open_connection(ip, port)
    with contextlib.closing(writer):
        while True:
            writer.write(message)
            await writer.drain()


if __name__ == "__main__":
    asyncio.run(flood('127.0.0.1', 8888))

```

tontinton committed:
bpo-41279: Add support for BufferedProtocol in _UnixReadPipeTransport (GH-21446)

The transport did not know how to use the proper api exported by
BufferedProtocol.

Added a new callback function that calls getbuffer() and
buffer_updated() instead of data_received() when the protocol given to
it is of type BufferedProtocol.

This is exactly the same way _SelectorSocketTransport handles a
BufferedProtocol.

tontinton committed:
bpo-41279: Add support for BufferedProtocol in _ProactorReadPipeTransport (GH-21446)

In the __init__ function if the protocol is of instance BufferedProtocol
instead of creating a buffer object, we call get_buffer on the protocol
to get its buffer.

In addition _loop_reading now calls _data_received as soon as there is
actual data instead of calling only after adding a recv_into event.

The reason for this change is because read_into could call it's callback
immediatly meaning overriding the data on the buffer before we actually
call _data_received on it, which fixes the potential issue of missed data.

1st1 commented on Lib/asyncio/proactor_events.py:
```diff
@@ -280,11 +287,13 @@ def _loop_reading(self, fut=None):
-
-                    data = self._data[:length]
+                    if length > -1:
+                        if not isinstance(self._protocol, protocols.BufferedProtocol):
```
Do we have proper functional tests that hit this branch? If not we need to add them.

1st1 replied:
You can probably copy some tests from the uvloop project (as I own the project there will be no licensing issues even if it's a literal copy)

tontinton replied:
I am thinking of doing parameterize on all the appropriate test in test_proactor_events.py.

That way we can use the already existing tests on both Protocol and BufferedProtocol

tontinton replied:
Ok I did this using the parameterize function, it did made me write a bit of ugly code I want you to look at.

Check out what I did in the test adding commit.

By the way shouldn't I add the same tests for ``UnixReadPipeTransportTests``?

tontinton replied:
Ok I added the tests to ``UnixReadPipeTransportTests`` as well using the same patches

1st1 replied:
> Ok I did this using the parameterize function, it did made me write a bit of ugly code I want you to look at.

Yeah, let's try to simplify it. See my other comment.

> By the way shouldn't I add the same tests for UnixReadPipeTransportTests?

I guess since we're updating `_UnixReadPipeTransport` we should.

tontinton replied:
Yeah I simplified it a lot

mpaolini commented on Lib/asyncio/proactor_events.py:
```diff
@@ -244,6 +248,17 @@ def _eof_received(self):
+        try:
+            self._protocol.buffer_updated(length)
+        except (SystemExit, KeyboardInterrupt):
+            raise
```
```suggestion
```
the two llines can be removed

tontinton replied:
Fixed

mpaolini commented on Lib/asyncio/proactor_events.py:
```diff
@@ -244,6 +248,15 @@ def _eof_received(self):
+            self._fatal_error(exc,
+                              'Fatal error: protocol.buffer_updated() '
+                              'call failed.')
+            return
```
```suggestion
```

just a small change ;)

tontinton replied:
Fixed

tzickel commented on Lib/asyncio/streams.py:
```diff
@@ -280,10 +283,16 @@ def connection_lost(self, exc):
-    def data_received(self, data):
+    def get_buffer(self, sizehint):
+        if self._buffer is None:
+            size = self._default_buffer_size if sizehint == -1 else sizehint
```
is sizehint is dynamic, this will keep it at a fixed size (which at the beginning might be small?)

tontinton replied:
You are right, I should probably ignore that argument.
Fixed.

1st1 commented on Lib/test/test_asyncio/test_unix_events.py:
```diff
@@ -642,12 +643,19 @@ def test_sock_sendfile_exception(self):
 
 
-class UnixReadPipeTransportTests(test_utils.TestCase):
+@parameterize
```
This is too much magic for a simple need. I suggest to factor out common tests and empty helpers into a base class and then use that to create two test cases, e.g.:

```python

class BaseTests:

    def make_test_protocol(self):
        raise NotImplementedError

    def test_something_1(self):
         ...
         ... self.make_test_protocol() ...

    def test_something_2(self):
         ...
         ... self.make_test_protocol() ...


class ProtocolTests(BaseTests, unittest.TestCase):

    def make_test_protocol(self):
        return asyncio.Protocol


class BufferedProtocolTests(BaseTests, unittest.TestCase):

    def make_test_protocol(self):
        return asyncio.BufferedProtocol
```

tontinton replied:
Ok I simplified it a whole lot now :)

1st1 commented on Lib/asyncio/unix_events.py:
```diff
@@ -488,7 +493,37 @@ def __repr__(self):
+                raise RuntimeError('get_buffer() returned an empty buffer')
+        except (SystemExit, KeyboardInterrupt):
+            raise
+        except BaseException as exc:
```
~Is there some specific need to handle BaseException here instead of Exception?~

Never mind, this is how we typically invoke callback methods.

tontinton replied:
👍

tontinton committed:
bpo-41279: Add BufferedProtocol to proactor transport tests (GH-21446)

tontinton committed:
bpo-41279: Add BufferedProtocol to unix read pipe transport tests (GH-21446)

tontinton committed:
bpo-41279: Support resetting a protocol in _ProactorReadPipeTransport (GH-21446)

When calling set_protocol to change the protocol you can now change the
type of the protocol from BufferedProtocol to Protocol or vice versa.

start_tls needed this feature as it could read into a buffered protocol
at first and then change the protocol to SSLProto which is a regular
protocol.

1st1 commented on Doc/library/asyncio-stream.rst:
```diff
@@ -65,6 +66,14 @@ and work with streams:
    returned :class:`StreamReader` instance.  By default the *limit*
    is set to 64 KiB.
+   when the *limit* is 0 allocate a global buffer to read into.
+   By allocating a global buffer, each read call will read directly into the
```
IMO this is a very low-level optimization for a very high-level (and frozen!) API. -1 on this.

tontinton replied:
I think that the optimization allows for too good of a performance boost to pass on.
If you have a suggestion on how we could maybe have this feature without changing the api it would be nice.
If not, maybe we can just add the ``StreamReaderBufferedProtocol`` but not add the api, as I want to use this feature.

tontinton replied:
Ok I updated the PR to not have this API change

1st1 replied:
> I think that the optimization allows for too good of a performance boost to pass on.

Have you benchmarked this specific optimization? If so please attach results, otherwise this is very speculative.

tontinton replied:
I a did benchmark using the scripts I wrote at the description of this PR.
It's probably not the best benchmark ever but it does provide me some good results.
Also it's not that far fetched that not calling malloc / free each time you read a buffer improves performance.

1st1 replied:
Huh, I didn't notice the benchmark results in the opening message.

>  It's probably not the best benchmark ever but it does provide me some good results.

You can potentially have a deque of 64kb buffers. Every stream could grab a buffer or allocate a new one if the deque is empty. The deque could have a max len set to 4 or some other relatively low number. That way you can have the benefits of global buffer and no visible API changes.

tontinton replied:
I thought about that, the issue is that the only doing this right now is giving for example the first 4 protocols to be of type ``StreamReaderBufferedProtocol``, but there is no way to update that one these protocols got closed unless I implement __del__ on that class, which I don't think is the best way to do this.

If I'll come up with some better idea, I'll either update this PR or work on a new one.

tontinton committed:
bpo-41279: Add StreamReaderBufferedProtocol (GH-21446)

This class gets better performance as BufferedProtocol uses read_into
into a buffer allocated before, instead of allocating a new buffer
each time read is called.

1st1 commented on Lib/asyncio/proactor_events.py:
```diff
@@ -218,12 +237,20 @@ def resume_reading(self):
-            # Call the protocol methode after calling _loop_reading(),
-            # since the protocol can decide to pause reading again.
-            self._loop.call_soon(self._data_received, self._data[:length], length)
+        if isinstance(self._protocol, protocols.BufferedProtocol):
```
I don't see the corresponding changes in selector_events.py. Does it handle these use cases already?

tontinton replied:
Yes it has the following bit of code:
```python
    def set_protocol(self, protocol):
        if isinstance(protocol, protocols.BufferedProtocol):
            self._read_ready_cb = self._read_ready__get_buffer
        else:
            self._read_ready_cb = self._read_ready__data_received

        super().set_protocol(protocol)
```

1st1 reviewed (commented):
This is one high quality PR, thanks for working on it. The fixes in proactor_events.py alone are great. I've left a couple comments.

tzickel commented on Lib/asyncio/proactor_events.py:
```diff
@@ -244,7 +271,15 @@ def _eof_received(self):
+                              'Fatal error: protocol.buffer_updated() '
+                              'call failed.')
+
+    def _buffer_updated(self, length):
```
bug, this function is defined twice now

tontinton replied:
Well it worked but just defined twice,
Fixed.

asvetlov commented on Lib/asyncio/streams.py:
```diff
@@ -307,6 +302,30 @@ def __del__(self):
+            reader.feed_data(data)
+
+
+class StreamReaderBufferedProtocol(BaseStreamReaderProtocol, protocols.BufferedProtocol):
```
Sorry, I don't follow.
Where is `StreamReaderBufferedProtocol` class used (except tests)?

tontinton replied:
It isn't used right now, as there is no way to include this class without changing or adding to the API.
It will be useful in the future though, once we do make API changes.

asvetlov replied:
Please elaborate.
From my understanding, the PR contains dead code only. Despite the code correctness, it is not used by the implementation of the streams.
What future API changes are you talking about? Can we apply these changes as a part of this pull request?

I like this PR, it makes a value if the buffered protocol is used. But if this protocol is not used -- why we need the PR at all?

I can imagine a little different approach: don't extract `BaseStreamReaderProtocol` but inherit `StreamReaderProtocol` from both `Protocol` and `BufferedProtocol`, implementing all `data_received()`, `get_buffer()` and `update_buffer()` in the same class.
asyncio selects the best approach currently: it uses the buffered protocol version if the protocol is an instance of `BufferedProtocol` with fallback to `data_received()` otherwise. We can utilize this facility.

What do you think?

tontinton replied:
The problem with this approach is that the `BufferedProtocol` allocates a 64k byte buffer, so on servers with a lot of connections, memory will start running out quickly as each stream now holds a 64k buffer.

For example on a server with 10k clients, python will hold ~640mb of memory.

This is why we should probably make it a custom choice of when to use the new `BufferedProtocol` **or** creating some kind of a protocol pool of a small size, so that the first few streams will be faster.

But this means that we either change / add API or choose a set number defaulted to the protocol pool.

Btw, doing the buffered protocol pool approach is not really trivial as I don't currently know when a `BufferedProtocol` needs to return to the pool (maybe I can think of something if you tell me that it's worth it, I'll be happy to implement something like that)

asvetlov replied:
We already have `limit` argument to control *output* buffer.
We can reuse the same number for controlling *input* buffer size as well.
In practice, I very doubt if people need different values for that.

I can imagine a variadic input buffer size starting from some default value and growing up to a limit but I doubt if we need it really. A simple implementation can be good enough.

A pool of buffers is an interesting idea but, again, the feature can be added later.

Thoughts?

tontinton replied:
> We already have limit argument to control output buffer.
> We can reuse the same number for controlling input buffer size as well.
> In practice, I very doubt if people need different values for that.

I don't really understand what you mean by using the ``limit`` argument as an *input* buffer size, as ``StreamReader``'s ``limit`` argument actually controls the buffer limit of what the ``read`` apis (such as ``read_exactly`` *or* ``read_line``) use to get the buffer from, so really this argument controls the *input* buffer size.

At the beginning what I did was that if ``limit`` received was 0, then I would use the ``BufferedProtocol`` instead of the regular ``Protocol``, but @1st1 didn't like this as it was essentially an API change (which is frozen, understandably).

> I can imagine a variadic input buffer size starting from some default value and growing up to a limit but I doubt if we need it really. A simple implementation can be good enough.

I believe what you mean is already implemented in ``StreamReader`` but I am not sure.
Also I am against reallocating the buffer as it would remove all benefits of ``BufferedProtocol``.

> A pool of buffers is an interesting idea but, again, the feature can be added later.

I think this is the best solution, though we need to think if we want that feature to be in this PR or on a different one.

By the way, if you know of a nice way of knowing when we should return a ``BufferedProtocol`` to the pool it would be nice to know your thoughts.
Right now I am thinking of creating a callback that is called when the protocol has closed or something like that.

msoxzw replied:
I reckon that the default buffer size is not necessarily decided by `limit` argument. We might start with a relatively small buffer, e.g. `io.DEFAULT_BUFFER_SIZE` or less, then expand and shrink buffer on demand.

Additionally, it appears that `feed_data` inflicts extra copy, so is it viable to directly write the buffer in `BaseStreamReaderProtocol` instead of the buffer in `StreamReaderBufferedProtocol`, or more aggresively utilize efficient and mature [`io.BufferedIOBase`](https://docs.python.org/3/library/io.html#io.BufferedIOBase) to implement `StreamReaderBufferedProtocol`?

gst commented on Lib/test/test_asyncio/test_proactor_events.py:
```diff
@@ -489,6 +407,191 @@ def test_dont_pause_writing(self):
+        self.protocol.data_received.assert_called_with(msgs[1])
+
+        tr.resume_reading()
+        tr.resume_reading()
```
why this double `pause_reading()` above and then `resume_reading()` here ?

and same question for below in `test_pause_resume_reading`.

tontinton replied:
Fixed

gst commented on Lib/test/test_asyncio/test_proactor_events.py:
```diff
@@ -489,6 +407,191 @@ def test_dont_pause_writing(self):
+
+        for msg in msgs[:2]:
+            self.loop._run_once()
+            self.protocol.data_received.assert_called_with(msg)
```
before the `_run_once()` would not it be better to assert that the `msg` is not in the `_mock_call_args_list` of the mock object ?

and same below for the others `data_received.assert_called_with` following a `_run_once()`.

tontinton replied:
I don't understand which mock object you are talking about, and I don't see any other test reading `_mock_call_args_list`.

Can you elaborate?

gst replied:
something like:

```python
for msg in msgs[:2]:
            self.assertFalse(unittest.mock.call(msg) in self.protocol.data_received.mock_calls)
            self.loop._run_once()
            self.protocol.data_received.assert_called_with(msg)
```

to ensure that each `msg` is effectively processed during the _run_once() call.

tontinton replied:
Oh I see, but protocol's `buffer_updated` (data_received is only in a regular stream object, not buffered) is not a mock, it's an actual function, so no `mock_calls` field.

gst replied:
ah ok :/
